common/djangoapps/student/models.py

```diff
@@ -1293,6 +1293,23 @@ class Meta:  # pylint: disable=missing-docstring
     def __unicode__(self):
         return "[CourseEnrollmentAllowed] %s: %s (%s)" % (self.email, self.course_id, self.created)
 
+    @classmethod
+    def may_enroll_and_unenrolled(cls, course_id):
+        """
+        Return QuerySet of students who are allowed to enroll in a course.
+
+        Result excludes students who have already enrolled in the
+        course.
+
+        `course_id` identifies the course for which to compute the QuerySet.
+        """
+        enrolled = CourseEnrollment.users_enrolled_in(course_id)
+        may_enroll = CourseEnrollmentAllowed.objects.filter(course_id=course_id)
+        return (
+            student for student in may_enroll if not
+            enrolled.filter(email=student.email).exists()
```

Comment:
This is going to require a bunch of database queries. Make use of a flat value set to simplify this. https://docs.djangoproject.com/en/1.4/ref/models/querysets/#values-list